File: src/main/java/de/axelrindle/broadcasterplugin/Broadcaster.java
package de.axelrindle.broadcasterplugin;

import com.google.common.io.Files;
import org.apache.commons.io.IOUtils;
import org.bukkit.Bukkit;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.plugin.java.JavaPlugin;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.List;
import java.util.logging.Logger;

/**
 * Main class of the Broadcaster plugin.
 */
@SuppressWarnings({"WeakerAccess", "ResultOfMethodCallIgnored"})
public final class Broadcaster extends JavaPlugin {

    private static final String consolePrefix = "[Broadcaster] ";
    public static final String prefix = "&2Broadcaster &f> ";
	public final Logger log = Logger.getLogger("Broadcaster");

	private FileConfiguration config;

	public List<String> messages;
	
	@Override
	public void onEnable() {
		log.info(consolePrefix + "Loading...");

		log.info(consolePrefix + "Loading config...");
        try {
            loadConfig();
        } catch (IOException e) {
            log.severe("Failed to load config! Please check your configuration!");
            e.printStackTrace();
            Bukkit.getPluginManager().disablePlugin(this);
            return;
        }
        log.info(consolePrefix + "Config loaded.");

        log.info(consolePrefix + "Loading messages...");
        try {
            loadMessages();
        } catch (IOException e) {
            log.severe(consolePrefix + "Failed to load messages! Please check your configuration!");
            e.printStackTrace();
            Bukkit.getPluginManager().disablePlugin(this);
            return;
        }
        log.info(consolePrefix + "Messages loaded.");

        if(getConfig().getBoolean("Cast.OnServerStart")) {
			startBroadcast();
		}

        getCommand("brc").setExecutor(new BrcCommand(this));
		
		log.info(consolePrefix + "Done! Version " + getDescription().getVersion());
	}
	
	@Override
	public void onDisable() {
		stopBroadcast();
		
		log.info(consolePrefix + "Successfully disabled!");
	}
	
	private void startBroadcast() {
		int interval = getConfig().getInt("Cast.Interval");
		BroadcastThread.start(this, messages, interval);
	}
	
	private void stopBroadcast() {
	    BroadcastThread.stop();
	}

    @Override
    public void reloadConfig() {
        log.info(consolePrefix + "Reloading config and messages...");
        try {
            loadConfig();
            loadMessages();
        } catch (IOException e) {
            log.severe(consolePrefix + "Failed to reload config! Please check your configuration!");
            e.printStackTrace();
            Bukkit.getPluginManager().disablePlugin(this);
            return;
        }
        log.info(consolePrefix + "Config and messages reloaded.");
    }

    @Override
	public FileConfiguration getConfig() {
		return config;
	}

	private void loadConfig() throws IOException {
	    File file = new File("plugins/Broadcaster/config.yml");
	    if(!file.exists()) {
            Files.createParentDirs(file);
            file.createNewFile();
            IOUtils.copy(getResource("config.yml"), new FileOutputStream(file));
        }

	    config = YamlConfiguration.loadConfiguration(file);
  	}
  	
  	private void loadMessages() throws IOException {
        File file = new File("plugins/Broadcaster/messages.yml");
        if(!file.exists()) {
            Files.createParentDirs(file);
            file.createNewFile();
            IOUtils.copy(getResource("messages.yml"), new FileOutputStream(file));
        }

        FileConfiguration config = YamlConfiguration.loadConfiguration(file);
        messages = config.getStringList("Messages");
    }
}

File: src/main/kotlin/de/axelrindle/broadcaster/Broadcaster.kt
package de.axelrindle.broadcaster

import de.axelrindle.broadcaster.command.BrcCommand
import de.axelrindle.pocketknife.PocketCommand
import de.axelrindle.pocketknife.PocketConfig
import de.axelrindle.pocketknife.PocketLang
import net.milkbowl.vault.permission.Permission
import org.apache.commons.io.IOUtils
import org.bukkit.Bukkit
import org.bukkit.plugin.PluginDescriptionFile
import org.bukkit.plugin.RegisteredServiceProvider
import org.bukkit.plugin.java.JavaPlugin
import org.bukkit.plugin.java.JavaPluginLoader
import java.io.*
import java.nio.charset.StandardCharsets

/**
 * The main plugin class. Does initialization and config loading.
 */
class Broadcaster : JavaPlugin {

    constructor() : super()

    // Required for mocking
    constructor(loader: JavaPluginLoader, description: PluginDescriptionFile, dataFolder: File, file: File?) :
        super(
            loader,
            description,
            dataFolder,
            file ?: File(Broadcaster::class.java.protectionDomain.codeSource.location.path)
        )

    companion object {
        const val CHAT_PREFIX = "&2Broadcaster &f> &r"
        private var instance: Broadcaster? = null

        fun get() = instance!!
    }

    internal val config = PocketConfig(this)
    internal val localization = PocketLang(this, config)

    internal var hasPluginPlaceholderApi: Boolean = false
        private set
    internal var hasPluginVault: Boolean = false
        private set

    private var _permission: Permission? = null
    internal val permission: Permission
        get() = _permission!!

    override fun onEnable() {
        logger.info("Startup...")

        // instance setup
        if (instance == null) {
            instance = this
        } else {
            logger.severe("An instance already exists!?")
            Bukkit.getPluginManager().disablePlugin(this)
            return
        }

        // loading configuration files
        logger.info("Loading configuration...")
        try {
            config.register("config", getResource("config.yml")!!)
            config.register("messages", getResource("messages.yml")!!)

            // manually create example json file
            createJsonExample()
        } catch (e: IOException) {
            logger.severe("Failed to load configuration files!")
            logger.severe(e.message)
            Bukkit.getPluginManager().disablePlugin(this)
            return
        }

        // localization init
        logger.info("Loading localization...")
        localization.addLanguages("en", "de")
        localization.init()

        logger.info("Initializing main functionality...")

        BroadcastingThread.loadMessages()
        logger.info("Loaded " + BroadcastingThread.messages.size + " messages.")

        // register command
        PocketCommand.register(this, BrcCommand())

        // check soft dependencies
        if (Bukkit.getPluginManager().getPlugin("PlaceholderAPI") != null) {
            hasPluginPlaceholderApi = true
            logger.info("Connected with PlaceholderAPI.")
        }
        if (setupPermissions()) {
            hasPluginVault = true
            logger.info("Connected with VaultAPI.")
        }

        // start casting if we should
        val mainConfig = config.access("config")!!
        val pauseOnEmptyServer = mainConfig.getBoolean("Cast.PauseOnEmptyServer")

        if (pauseOnEmptyServer) {
            Bukkit.getPluginManager().registerEvents(BroadcastingThread.EventListener(), this)
        }

        if (mainConfig.getBoolean("Cast.OnServerStart")) {

            // if the broadcast should be paused on empty server, start with an initial pause,
            // because no player will be online immediately after server start
            if (pauseOnEmptyServer) {
                BroadcastingThread.paused = true
            }

            // else just start the broadcast normally
            else {
                BroadcastingThread.start()
            }
        }

        logger.info("Done! v${description.version}")
    }

    private fun createJsonExample() {
        val indicatorFile = File(dataFolder, "examples_created")
        if (indicatorFile.exists().not()) {
            var success = true

            val jsonDir = File(dataFolder, "json")
            success = success && jsonDir.mkdir()

            val jsonFile = File(jsonDir, "visit-github.json")
            success = success && jsonFile.createNewFile()
            FileOutputStream(jsonFile).use { fos ->
                javaClass.getResourceAsStream("/json/visit-github.json").use { fis ->
                    fis.copyTo(fos)
                }
            }

            success = success && indicatorFile.createNewFile()
            FileOutputStream(indicatorFile).use { fos ->
                IOUtils.write("This file indicates that example files have been created " +
                        "and do not need to be created again. For now, this only includes " +
                        "the json/ directory.", fos, StandardCharsets.UTF_8)
            }

            if (success) {
                logger.info("Created example file at ${jsonFile.absolutePath}.")
            } else {
                logger.warning("Failed to create example file!")
            }
        }
    }

    private fun setupPermissions(): Boolean {
        val rsp: RegisteredServiceProvider<Permission>? = server.servicesManager.getRegistration(Permission::class.java)
        _permission = rsp?.provider
        return _permission != null
    }

    override fun onDisable() {
        try {
            BroadcastingThread.stop()
        } catch (e: RuntimeException) {
            // ignore
        }
    }
}